hashing/ripemd_160.cpp

```diff
@@ -0,0 +1,437 @@
+ */
+
+#include <cassert>   /// For assert
+#include <cstdint>   /// For data types such as std::int32_t, std::uint32_t, etc
+#include <iomanip>   /// For functions like std::setw, std::setfill
+#include <iostream>  /// For managing io
+#include <sstream>   /// For bytes to hex string
+#include <stdexcept>  /// For standard exceptions like std::invalid_argument
+#include <string>     /// For string data
+#include <thread>     /// To parallelize for efficiency
+#include <vector>     /// For dynamic arrays
+
+/**
+ * @namespace hashing
+ * @brief Hashing algorithms
+ */
+namespace hashing {
+/**
+ * @class RIPEMD-160
+ * @brief Implementation of relevant functions to perform RIPEMD160 hashing.
+ * This class expects a string to hash.
+ */
+class RIPEMD160 {
+ private:
+    /**
+     * @brief Implements f(j,B,C,D)
+     * @param j Round number j / 16
+     * @param B,C,D The state values
+     * @return Returns the function value
+     * @note 0 <= j <= 4
+     */
+    uint32_t f(int j, uint32_t B, uint32_t C, uint32_t D) {
+        switch (j) {
+            case 0:
+                return B ^ C ^ D;
+            case 1:
+                return (B & C) | (~B & D);
+            case 2:
+                return (B | ~C) ^ D;
+            case 3:
+                return (B & D) | (C & ~D);
+            case 4:
+                return B ^ (C | ~D);
+            default:
+                throw std::invalid_argument("j value out of bound");
+        }
+    }
+
+    /**
+     * @brief Implements K value for a given j
+     * @param j Round number j / 16
+     * @return Appropriate K value
+     * @note 0 <= j <= 4
+     */
+    uint32_t K(int j) {
+        switch (j) {
+            case 0:
+                return static_cast<uint32_t>(0x00000000);
+            case 1:
+                return static_cast<uint32_t>(0x5A827999);
+            case 2:
+                return static_cast<uint32_t>(0x6ED9EBA1);
+            case 3:
+                return static_cast<uint32_t>(0x8F1BBCDC);
+            case 4:
+                return static_cast<uint32_t>(0xA953FD4E);
+            default:
+                throw std::invalid_argument("j value out of bound");
+        }
+    }
+
+    /**
+     * @brief Implements K' value for a given j
+     * @param j Round number j / 16
+     * @return Appropriate K' value
+     * @note 0 <= j <= 4
+     */
+    uint32_t K_dash(int j) {
+        switch (j) {
+            case 0:
+                return 0x50A28BE6;
+            case 1:
+                return 0x5C4DD124;
+            case 2:
+                return 0x6D703EF3;
+            case 3:
+                return 0x7A6D76E9;
+            case 4:
+                return 0x00000000;
+            default:
+                throw std::invalid_argument("j value out of bound");
+        }
+    }
+
+    /**
+     * @brief Specifies r value for a given j.
+     *
+     * @details Specifies the order in which 4-byte words of the current
```

Comment:
these could be merged into the main documentation with a heading on tables
```
/**
 * @file
 ...
 * ### tables
 * table in the class are used for ...
 * they specify the order ...
```

Reply:
suggestions updated